### PR TITLE
Update schema for currency swap

### PR DESCRIPTION
**Pull request by sdavidp**

### CODE REVIEW

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (AlexBrandes) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:93]: In the current case, this would accept two SpendTransaction instances. 

The ideal implementation would be to make this a wrapper (however implemented) that could accept different transaction types so that it can be extended in the future rather than being fixed to this specific case.

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (AlexBrandes) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:86]: Let's make sure this field's name matches the corresponding field in AllowSpend, making both either `lastValidOrdinal` or `validUntilOrdinal`.

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (sdavidp) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala]: removed

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (sdavidp) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala]: Adjusted class name, field types, and extended a trait.

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (sdavidp) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala]: changed. also adjusted `currencyId` to match.

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:21]: How the `CurrencyId` may have an empty `value`? Isn't it non-existent `CurrencyId`?
IMO it should be an `Option[CurrencyId]` in the `AllowSpend` class if the currencyId is an optional type

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:37]: just `approvers`?

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:32]: IMO we shouldn't mix it with `Transaction` class, so `SwapAmount` and `SwapSalt` in here 🤔

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:61]: IMO there should be no differentiation for transaction type. We already had `DAGTransaction` and `CurrencyTransaction` that we decided to merge under just `Transaction` and use composition over inheritance to extend the functionalities. The same should be applied to swaps

Comment (AlexBrandes) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:61]: There are some pretty significant differences between SpendTransaction and Transaction. The most important one imo is that SpendTransaction is Signed by the approver(s) of an AllowSpend, not the address transferring the tokens. It also has a required reference to an allowSpend which is needed for validation of the `amount` field. Do you think those aspects would still be best supported as Transaction w/some metadata?

Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:61]: > The most important one imo is that SpendTransaction is Signed by the approver(s) of an AllowSpend, not the address transferring the tokens

The `Transaction` itself is just a data class. We have a separate `Signed[A]` wrapper and the signing process can be defined differently for various purposes. What you wrote is just the validation function -> we would require this or that signature. Anyway, it's disjoint

> It also has a required reference to an allowSpend which is needed for validation of the amount field.

This can be a metadata part indeed.

---

If the SpendTransaction should be handled and managed differently than a normal Transaction, then it's okay to make it a totally separate thing. If the only difference is some additional validation based on the metadata, I see no point for making it separate.

Comment (AlexBrandes) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:61]: Seems reasonable. I'll let others comment on it but I don't have an issue with that structure.

Comment (sdavidp) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:61]: @ryle-goehausen ?

Comment (sdavidp) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:61]: @marcinwadon @AlexBrandes @ryle-goehausen 

Are there any new/updated thoughts on this?

I see Marcin's concern about previously merging `DAGTransaction` and `CurrencyTransaction`. Isn't `SpendTransaction` a new/different thing? Isn't the flow going to be entirely different?

It feels to me that adding metadata to `Transaction` instead of creating a new class may introduce problems now or in the future with trying to assign meaning to existing `Transaction` fields that don't really fit or don't exist for `SpendTransaction`. E.g.  how do we set `Transaction.parent` ?

Comment (ryle-goehausen) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:67]: As discussed in meeting, I think it doesn't make sense to re-use the old transaction class. We can discuss creating a more generic transaction in the future, but the previous schema is locked and we want to migrate away from that. For now, we should follow updates to this by handling it as a special case and then discuss a way to handle this more generically in the future.

**Comment by marcinwadon** [comment]
please add a label

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:31]: is it a `TODO`? IMO you should create a ticket for that as it can easily get lost. otherwise, is the PR ready to be merged?

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:41]: is the `salt` required?

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:55]: as above

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:59]: as above

Comment (marcinwadon) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:59]: btw. I comment my code using `TODO: @mwadon lorem ipsum` so I can easily find my todos and resolve them before merging the PR

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/schema/swap.scala:70]: so if there is no currency defined, it is a DAG spend transaction, correct?

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (ryle-goehausen) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala]: Empty should indicate DAG, either level of option representation should convey the same thing, either it's a Some(None) or None for DAG indicator

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (ryle-goehausen) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:31]: I think we should just use the same fields as transaction reference here.

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (ryle-goehausen) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:41]: I think this was just a convention following the original transaction schema, I don't think it's really required there either due to unique transaction references. Not sure the original reason behind that.

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (ryle-goehausen) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:55]: These values will only be used in the metagraph itself, so it shouldn't matter what they are here, they'll be removed from tessellation and put in project marcus is creating.

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (ryle-goehausen) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:70]: yes

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (sdavidp) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:59]: removed comment

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (sdavidp) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:41]: removed

File: modules/shared/src/main/scala/org/tessellation/schema/swap.scala
Comment (sdavidp) [reply modules/shared/src/main/scala/org/tessellation/schema/swap.scala:31]: changed

**Review by marcinwadon — APPROVED** [review]
The schema looks okay for me

**Pull request merged**
